src/fec_base.h

```diff
@@ -603,6 +626,135 @@ bool FecCode<T>::decode_bufs(
     return true;
 }
 
+/**
+ * Perform a Lagrange interpolation to find the coefficients of the
+ * polynomial
+ *
+ * @note If all fragments are available ifft(words) is enough
+ *
+ * @param output must be exactly n_data
+ * @param props special values dictionary must be exactly n_data
+ * @param offset used to locate special values
+ * @param fragments_ids unused
+ * @param words v=(v_0, v_1, ..., v_k-1) k must be exactly n_data
+ */
+template <typename T>
+void FecCode<T>::decode(
+    vec::Vector<T>* output,
+    const std::vector<Properties>& props,
+    off_t offset,
+    vec::Vector<T>* fragments_ids,
+    vec::Vector<T>* words)
+{
+    int vx_zero;
+    // vector x=(x_0, x_1, ..., x_k-1)
+    vec::Vector<T> vx(this->gf, this->n_data);
+
+    // prepare for decoding
+    decode_prepare(props, offset, fragments_ids, words, &vx, &vx_zero);
+
+    // Lagrange interpolation
+    decode_Lagrange(output, props, offset, fragments_ids, words, &vx, vx_zero);
+}
+
+/* Prepare for decoding
+ * It supports for FEC using multiplicative FFT over FNT
+ */
+template <typename T>
+void FecCode<T>::decode_prepare(
+    const std::vector<Properties>& props,
+    off_t offset,
+    vec::Vector<T>* fragments_ids,
+    vec::Vector<T>* words,
+    vec::Vector<T>* vx,
+    int* vx_zero)
+{
+    int k = this->n_data; // number of fragments received
+    // vector x=(x_0, x_1, ..., x_k-1)
+    for (int i = 0; i < k; i++) {
+        vx->set(i, this->gf->exp(r, fragments_ids->get(i)));
+    }
+
+    for (int i = 0; i < k; i++) {
+        const int j = fragments_ids->get(i);
+        auto data = props[j].get(ValueLocation(offset, j));
+
+        // Check if the symbol is a special case whick is marked by "@".
+        // Note: this check is necessary when word_size is not large enough to
+        // cover all symbols of the field.
+        // Following check is used for FFT over FNT where the single special
+        // case symbol equals card - 1
+        if (data && *data == "@") {
+            words->set(i, fft->get_gf()->card() - 1);
+        }
+    }
+}
+
+/* Lagrange interpolation
+ * It supports for FEC using multiplicative FFT over GF differing NF4
+ */
+template <typename T>
+void FecCode<T>::decode_Lagrange(
+    vec::Vector<T>* output,
+    const std::vector<Properties>& props,
+    off_t offset,
+    vec::Vector<T>* fragments_ids,
+    vec::Vector<T>* words,
+    vec::Vector<T>* vx,
+    int vx_zero)
+{
+    if (inv_r_powers == nullptr) {
+        throw LogicError("The pointer \"inv_r_powers\" must be initialized");
```

Comment:
very minor: you could use single quote ' or backtick `, that would sidestep the need to escape with \